### PR TITLE
feat(thirdpartypasswordless): add implementation for thirdpartypasswordless

### PR DESCRIPTION
**Pull request by porcellus**
## Summary of change

Adding a combined third-party + passwordless recipe

## Related issues

-   supertokens/supertokens-core#330

## Test Plan

Adding tests later

## Documentation changes

Adding new entirely new recipe docs.

## Checklist for important updates

-   [x] Changelog has been updated
-   [x] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [x] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [x] Had run `npm run build-pretty`
-   [x] Had installed and ran the pre-commit hook
-   [x] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [x] If added a new recipe interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [x] If I added a new recipe, I also added the recipe entry point into the `size-limit` section of `package.json` with the size limit set to the current size rounded up.

## Remaining TODOs for this PR

-   [x] Docs
-   [x] New version?
-   [x] Tests
- [x] Resolve TODOs

### CODE REVIEW

File: lib/ts/recipe/passwordless/types.ts [hunk 1]
```diff
@@ -263,6 +263,7 @@ export type UserInput = (
           validatePhoneNumber?: (phoneNumber: string) => Promise<string | undefined> | string | undefined;
 
           signInUpFeature?: SignInUpFeatureConfigInput & {
+              defaultCountry?: CountryCode;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/passwordless/types.ts:266]: Is this a bug fix for passwordless recipe? This needs to go into another PR which we can release separately.

File: lib/ts/recipe/thirdpartypasswordless/types.ts [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    providers?: (Provider | CustomProviderConfig)[];
+};
+
+export type UserInput = (
+    | {
+          contactMethod: "EMAIL";
+
+          validateEmailAddress?: (email: string) => Promise<string | undefined> | string | undefined;
+
+          signInUpFeature?: SignInUpFeatureConfigInput;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/types.ts:80]: This is called `signInAndUpFeature` in thirdpartyemailpassword. But in `passwordless` recipe, it is called `signInUp`.. which one should we pick?

File: lib/build/recipe/thirdpartypasswordless/types.d.ts [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+        flowType: "USER_INPUT_CODE" | "MAGIC_LINK" | "USER_INPUT_CODE_AND_MAGIC_LINK";
+        lastResend: number;
+        redirectToPath?: string;
+    }) => Promise<void> | void;
+    clearLoginAttemptInfo: () => Promise<void> | void;
+    getOAuthAuthorisationURL: (input: { thirdPartyId: string; config: TPConfig }) => Promise<string>;
+    signInAndUp: (input: { thirdPartyId: string; config: TPConfig }) => Promise<
+        | {
+              status: "OK";
+              user: User;
```
Comment (rishabhpoddar) [inline lib/build/recipe/thirdpartypasswordless/types.d.ts:246]: This `User` type is not correct. It should be the same type as that in passwordless recipe (since that is a superset of third party user)

File: lib/build/recipe/thirdpartypasswordless/types.d.ts [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        preAuthSessionId: string;
+        contactInfo: string;
+        contactMethod: "EMAIL" | "PHONE";
+        flowType: "USER_INPUT_CODE" | "MAGIC_LINK" | "USER_INPUT_CODE_AND_MAGIC_LINK";
+        lastResend: number;
+        redirectToPath?: string;
+    }) => Promise<void> | void;
+    clearLoginAttemptInfo: () => Promise<void> | void;
+    getOAuthAuthorisationURL: (input: { thirdPartyId: string; config: TPConfig }) => Promise<string>;
+    signInAndUp: (input: { thirdPartyId: string; config: TPConfig }) => Promise<
```
Comment (rishabhpoddar) [inline lib/build/recipe/thirdpartypasswordless/types.d.ts:243]: Raname to `thirdPartySignInAndUp`?

File: lib/ts/recipe/thirdpartypasswordless/types.ts [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+    termsOfServiceLink?: string;
+    // TODO: better name
+    providerAndEmailOrPhoneFormStyle: Styles;
+    userInputCodeFormStyle: Styles;
+    linkSentScreenStyle: Styles;
+    closeTabScreenStyle: Styles;
+    disableDefaultImplementation?: boolean;
+};
+
+export type NormalisedConfig = {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/types.ts:142]: The NormalisedConfig should actually be the input type for passwordless and third-party - cause those will normalise the input in their recipe anyway.

File: package.json [hunk 6]
```diff
@@ -104,19 +104,23 @@
...
         },
         {
             "path": "recipe/thirdparty/index.js",
-            "limit": "100kb"
+            "limit": "110kb"
         },
         {
             "path": "recipe/emailpassword/index.js",
-            "limit": "100kb"
+            "limit": "110kb"
```
Comment (rishabhpoddar) [inline package.json:115]: Why this increase?

File: package.json [hunk 7]
```diff
@@ -104,19 +104,23 @@
         },
         {
             "path": "recipe/thirdpartyemailpassword/index.js",
-            "limit": "110kb"
+            "limit": "120kb"
         },
         {
             "path": "recipe/thirdparty/index.js",
-            "limit": "100kb"
+            "limit": "110kb"
```
Comment (rishabhpoddar) [inline package.json:111]: Why this increase?

File: package.json [hunk 8]
```diff
@@ -104,19 +104,23 @@
         },
         {
             "path": "recipe/thirdpartyemailpassword/index.js",
-            "limit": "110kb"
+            "limit": "120kb"
```
Comment (rishabhpoddar) [inline package.json:107]: Why this increase?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/thirdpartypasswordless/types.ts (hunk 2 above)
Comment (porcellus) [reply lib/ts/recipe/thirdpartypasswordless/types.ts:80]: Either is fine I think, I went with the one from passwordless because it uses more props from there.

File: package.json (hunk 7 above)
Comment (porcellus) [reply package.json:111]: Hmm, I'm not sure why but it was failing for me locally. I wanted to check if there was an increase in 0.18 and these weren't updated before finalizing this PR. Running it again now, it seems to be OK with the old limits as well. I'm reverting these.

File: lib/build/recipe/thirdpartypasswordless/types.d.ts (hunk 3 above)
Comment (porcellus) [reply lib/build/recipe/thirdpartypasswordless/types.d.ts:246]: I'm not sure. Why are we returning the superset when it's always the more specific type? That way devs using this would have to assert that the email prop is present on the returned value.

File: lib/ts/recipe/thirdpartypasswordless/types.ts (hunk 2 above)
Comment (rishabhpoddar) [reply lib/ts/recipe/thirdpartypasswordless/types.ts:80]: Okay!

File: lib/build/recipe/thirdpartypasswordless/types.d.ts (hunk 3 above)
Comment (rishabhpoddar) [reply lib/build/recipe/thirdpartypasswordless/types.d.ts:246]: Email may not be present if a user used passwordless via mobile. Hence it needs to be optional

File: lib/ts/recipe/thirdpartypasswordless/types.ts (hunk 5 above)
Comment (rishabhpoddar) [reply lib/ts/recipe/thirdpartypasswordless/types.ts:142]: @porcellus did you miss this?

File: lib/ts/recipe/thirdpartypasswordless/index.ts [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    static Apple = Apple;
+    static Facebook = Facebook;
+    static Github = Github;
+    static ThirdPartyPasswordlessAuth = ThirdPartyPasswordlessAuth;
+    static SignInAndUp = (prop?: any) =>
+        ThirdPartyPasswordless.getInstanceOrThrow().getFeatureComponent("signInUp", prop);
+    static SignInAndUpTheme = SignInUpTheme;
+    static EmailVerification = (prop?: any) =>
+        ThirdPartyPasswordless.getInstanceOrThrow().getFeatureComponent("emailverification", prop);
+    static EmailVerificationTheme = EmailVerificationTheme;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/index.ts:72]: Need to also add LinkClickedScreen

File: lib/ts/recipe/thirdpartypasswordless/recipe.tsx [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+            return (
+                <AuthWidgetWrapper<GetRedirectionURLContext, PreAPIHookContext, OnHandleEventContext, NormalisedConfig>
+                    authRecipe={this}
+                    history={props.history}>
+                    <SignInAndUp recipe={this} {...props} />
+                </AuthWidgetWrapper>
+            );
+        } else if (componentName === "linkClickedScreen") {
+            if (this.passwordlessRecipe === undefined) {
+                throw new Error("Should not come here...");
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/recipe.tsx:191]: Change the error message - it can come here in case the user wants to embed the link clicked screen, but they have disabled passwordless

File: lib/ts/recipe/thirdpartypasswordless/recipe.tsx [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+            );
+            return ThirdPartyPasswordless.instance;
+        };
+    }
+
+    static getInstanceOrThrow(): ThirdPartyPasswordless {
+        if (ThirdPartyPasswordless.instance === undefined) {
+            let error =
+                "No instance of ThirdPartyPasswordless found. Make sure to call the ThirdPartyPasswordless.init method." +
+                "See https://supertokens.io/docs/thirdpartyemailpassword/quick-setup/frontend";
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/recipe.tsx:229]: Change message to link to thirdpartypasswordless docs

File: lib/ts/recipe/thirdpartypasswordless/recipeImplementation/index.ts [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        consumeCode(input) {
+            return passwordlessImpl.consumeCode.bind(DerivedPWLess(this))(input);
+        },
+        createCode(input) {
+            return passwordlessImpl.createCode.bind(DerivedPWLess(this))(input);
+        },
+        doesPhoneNumberExist(input) {
+            return passwordlessImpl.doesPhoneNumberExist.bind(DerivedPWLess(this))(input);
+        },
+        doesEmailExist: async function (input: { email: string; config: EPConfig }) {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/recipeImplementation/index.ts:24]: Perhaps we should change this function name to `doesPasswordlessUserEmailExist`

File: lib/ts/recipe/thirdpartypasswordless/recipeImplementation/index.ts [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    const thirdPartyImpl = ThirdPartyRecipeImplementation(recipeId, appInfo);
+
+    return {
+        consumeCode(input) {
+            return passwordlessImpl.consumeCode.bind(DerivedPWLess(this))(input);
+        },
+        createCode(input) {
+            return passwordlessImpl.createCode.bind(DerivedPWLess(this))(input);
+        },
+        doesPhoneNumberExist(input) {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/recipeImplementation/index.ts:21]: Change name to `doesPasswordlessUserPhoneNumberExist`

File: lib/ts/recipe/thirdpartypasswordless/components/themes/signInUp/passwordlessForm.tsx [hunk 14]
```diff
@@ -0,0 +1,170 @@
+/* Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/components/themes/signInUp/passwordlessForm.tsx:1]: Why do we need to reimplement so much here? Let's discuss on a call.

File: lib/ts/recipe/thirdpartypasswordless/components/themes/signInUp/index.tsx [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+        return <ThirdPartyOnlySignInUpWrapper {...props} thirdPartyRecipe={props.thirdPartyRecipe} />;
+    }
+
+    return (
+        <PasswordlessSignInAndUp recipe={props.passwordlessRecipe} history={props.history} isEmbedded={true}>
+            <PasswordlessSignInAndUpForm
+                // Seed props. Real props will be given by parent feature.
+                {...({} as PasswordlessSignInUpThemeProps)}
+                header={props.thirdPartyRecipe === undefined ? <Header /> : <HeaderWithProviderList {...props} />}
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/components/themes/signInUp/index.tsx:48]: It's kind of strange that you are adding third party providers into the header of the passwordless form - as opposed to how it's being done in thirdpartyemailpassword. Maybe it's fine. But just wanna discuss on a call.

File: lib/ts/recipe/thirdpartypasswordless/components/themes/signInUp/thirdPartyOnlySignInUp.tsx [hunk 16]
```diff
@@ -0,0 +1,76 @@
+/* Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/components/themes/signInUp/thirdPartyOnlySignInUp.tsx:1]: Why is this file needed?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/thirdpartypasswordless/components/themes/signInUp/thirdPartyOnlySignInUp.tsx (hunk 16 above)
Comment (porcellus) [reply lib/ts/recipe/thirdpartypasswordless/components/themes/signInUp/thirdPartyOnlySignInUp.tsx:1]: This needed to be a separate component because it uses the style context as well as providing it and adding the keeping it in `index.ts` made it feel crowded. I just moved `HeaderWithProviderList` out of it as well.

File: lib/ts/recipe/thirdpartypasswordless/components/themes/signInUp/index.tsx (hunk 15 above)
Comment (porcellus) [reply lib/ts/recipe/thirdpartypasswordless/components/themes/signInUp/index.tsx:48]: It's passed from the outside because that way I don't have to handle the third-party recipe in the "re-implementation" of the passwordless sign-in form. Originally I solved this using a component override. We can decide which header to show inside the pwless form and pass the `thirdPartyRecipe` instead but that felt weird.

File: lib/ts/recipe/thirdpartypasswordless/components/themes/signInUp/passwordlessForm.tsx (hunk 14 above)
Comment (porcellus) [reply lib/ts/recipe/thirdpartypasswordless/components/themes/signInUp/passwordlessForm.tsx:1]: There are a few reasons:
- We need to override the header but it also depends on the state.
- We need to provide different style contexts based on the state
- We need to control how errors are shown.
I think there is a better solution that would also do away with the embedded feature components. Let's discuss this tomorrow.

File: lib/ts/recipe/thirdpartypasswordless/components/features/signInAndUp/index.tsx [hunk 17]
```diff
@@ -18,55 +18,146 @@
...
+            if (
+                pwlessState.error !== undefined &&
+                tpState.error !== undefined &&
+                tpState.error === "SOMETHING_WENT_WRONG_ERROR"
+            ) {
+                error = pwlessState.error;
+            }
+            return {
+                error,
+            };
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/components/features/signInAndUp/index.tsx:94]: Do add a comment explaining what's going on here.

File: lib/ts/recipe/thirdpartypasswordless/index.ts [hunk 18]
```diff
@@ -70,6 +70,9 @@ export default class Wrapper {
     static EmailVerification = (prop?: any) =>
         ThirdPartyPasswordless.getInstanceOrThrow().getFeatureComponent("emailverification", prop);
     static EmailVerificationTheme = EmailVerificationTheme;
+
+    static LinkClicked = (prop?: any) =>
+        ThirdPartyPasswordless.getInstanceOrThrow().getFeatureComponent("linkClickedScreen", prop);
 }
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/index.ts:76]: Need to add theme of this too

Comment (rishabhpoddar) [reply lib/ts/recipe/thirdpartypasswordless/index.ts:76]: Also, consider renaming it to `PasswordlessLinkClicked`?

File: lib/ts/recipe/thirdpartypasswordless/index.ts [hunk 19]
```diff
@@ -70,6 +70,9 @@ export default class Wrapper {
...
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/index.ts:77]: We should also expose ThirdPartyCallback screen in this, ThirdParty and ThirdPartyEmailPassword recipe?

File: lib/ts/recipe/thirdpartypasswordless/types.ts [hunk 20]
```diff
@@ -52,19 +63,21 @@ import { ComponentOverrideMap as PasswordlessOverrideMap } from "../passwordless
...
 
-export type ComponentOverrideMap = Omit<PasswordlessOverrideMap, "PasswordlessSignInUpHeader"> &
-    Omit<ThirdPartyOverrideMap, "ThirdPartySignUpFooter" | "ThirdPartySignUpHeader"> & {
-        ThirdPartyPasswordlessHeader?: ComponentOverride<typeof Header>;
+export type ComponentOverrideMap = Omit<PasswordlessOverrideMap, "PasswordlessSignInUpHeader_Override"> &
+    Omit<ThirdPartyOverrideMap, "ThirdPartySignUpFooter_Override" | "ThirdPartySignUpHeader_Override"> & {
+        ThirdPartyPasswordlessHeader_Override?: ComponentOverride<typeof Header>;
     };
 
 // TODO: move the definition here if necessary
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/types.ts:78]: TODO in comment.

File: lib/ts/recipe/thirdpartypasswordless/types.ts [hunk 21]
```diff
@@ -121,38 +134,12 @@ export type UserInput = (
...
 export type NormalisedConfig = {
-    contactMethod: "PHONE" | "EMAIL" | "EMAIL_OR_PHONE";
-
-    validateEmailAddress: (email: string) => Promise<string | undefined> | string | undefined;
-    validatePhoneNumber: (phoneNumber: string) => Promise<string | undefined> | string | undefined;
+    passwordlessUserInput: PwlessUserInput | undefined;
+    thirdpartyUserInput: TPUserInput | undefined;
 
-    signInUpFeature: NormalizedSignInUpFeatureConfig;
+    providerAndEmailOrPhoneFormStyle: Styles | undefined;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/types.ts:141]:
```suggestion
    thirdPartyProviderAndEmailOrPhoneFormStyle: Styles | undefined;
```

File: lib/ts/recipe/thirdpartypasswordless/types.ts [hunk 22]
```diff
@@ -169,12 +156,50 @@ export type PreAPIHookContext = PasswordlessPreAPIHookContext | ThirdPartyPreAPI
...
+    | {
+          activeScreen: SignInUpScreens;
+          passwordlessRecipe: PWlessRecipe;
+          pwlessState: PWlessSignInUpState;
+          pwlessDispatch: Dispatch<PasswordlessSignInUpAction>;
+          pwlessChildProps: PwlessChildProps;
+      }
+);
+
 export type TPPWlessRecipeInterface = {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/types.ts:203]: In this, consider renaming:
- `getLoginAttemptInfo` -> `getPasswordlessLoginAttemptInfo`
- `setLoginAttemptInfo` -> `setPasswordlessLoginAttemptInfo`

File: lib/ts/recipe/thirdpartypasswordless/utils.ts [hunk 23]
```diff
@@ -16,30 +16,17 @@
...
-        (config.signInUpFeature === undefined ||
-            config.signInUpFeature.providers === undefined ||
-            config.signInUpFeature.providers.length === 0)
-    ) {
+    const disableThirdParty =
+        config.signInUpFeature === undefined ||
+        config.signInUpFeature.providers === undefined ||
+        config.signInUpFeature.providers.length === 0;
+    if (disablePasswordless && disableThirdParty) {
         throw new Error("You need to enable either email password or third party providers login.");
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdpartypasswordless/utils.ts:30]: Change error message

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/build/recipe/thirdpartypasswordless/types.d.ts (hunk 3 above)
Comment (rishabhpoddar) [reply lib/build/recipe/thirdpartypasswordless/types.d.ts:275]: @porcellus did you miss this?

Comment (rishabhpoddar) [reply lib/build/recipe/thirdpartypasswordless/types.d.ts:275]: In a recipe, we use just one user type across all of its functions. In this recipe, the user type should have both email and phone number as optional, therefore, this function must return that user type.

Comment (porcellus) [reply lib/build/recipe/thirdpartypasswordless/types.d.ts:275]: One problem is that we need to assign this to the signInAndUp of the third-party recipe, which requires this user type. Other than that, maybe `PasswordlessUser | User` would be the most precise because that is what the user receives in the `SUCCESS` event.

File: CHANGELOG.md [hunk 24]
```diff
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html)
 
+## [unreleased]
+
+### Added
+
+-   ThirdParty+Passwordless recipe
+
```
[the remark this answers is not in the document]
Comment (porcellus) [reply CHANGELOG.md:13]: I reverted the bugfix part because I later discovered that it could cause some problems.

[the remark this answers is not in the document]
Comment (rishabhpoddar) [reply CHANGELOG.md:13]: But isn't the bug fix needed anyway? I'm a bit confised.

File: examples/with-thirdpartypasswordless/pm2.json [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+            "name": "server",
+            "exec_mode": "cluster",
+            "instances": 1,
+            "watch": true,
+            "log_date_format": "YYYY-MM-DD HH:mm Z",
+            "merge_logs": true,
+            "env": {
+                "NODE_ENV": "production",
+                "REACT_APP_API_PORT": "10001",
+                "REACT_APP_API_URL": "https://api-passwordless.demo.supertokens.com",
```
Comment (rishabhpoddar) [inline examples/with-thirdpartypasswordless/pm2.json:14]: Need to change link to api-thirdpartypasswordless.demo.supertokens.com

File: examples/with-thirdpartypasswordless/pm2.json [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+                "REACT_APP_WEBSITE_PORT": "10000",
+                "REACT_APP_WEBSITE_URL": "https://passwordless.demo.supertokens.com"
```
Comment (rishabhpoddar) [inline examples/with-thirdpartypasswordless/pm2.json:16]: Need to change link to thirdpartypasswordless.demo.supertokens.com

File: examples/with-thirdpartypasswordless/pm2.json [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+            "name": "api-server",
+            "exec_mode": "cluster",
+            "instances": 1,
+            "watch": true,
+            "log_date_format": "YYYY-MM-DD HH:mm Z",
+            "merge_logs": true,
+            "env": {
+                "NODE_ENV": "production",
+                "REACT_APP_API_PORT": "10001",
+                "REACT_APP_API_URL": "https://api-passwordless.demo.supertokens.com",
```
Comment (rishabhpoddar) [inline examples/with-thirdpartypasswordless/pm2.json:30]: Need to change link to api-thirdpartypasswordless.demo.supertokens.com

File: examples/with-thirdpartypasswordless/pm2.json [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+                "REACT_APP_WEBSITE_PORT": "10000",
+                "REACT_APP_WEBSITE_URL": "https://passwordless.demo.supertokens.com"
```
Comment (rishabhpoddar) [inline examples/with-thirdpartypasswordless/pm2.json:32]: Need to change link to thirdpartypasswordless.demo.supertokens.com

File: examples/with-thirdpartypasswordless/scripts/start_container.sh [hunk 29]
```diff
@@ -0,0 +1,10 @@
+docker run -d \
+    --restart=always \
+    --name demo-app-passwordless \
+    --label name=demo-app \
+    --label type=app \
+    --label mode=production \
+    --volume /home/ubuntu/supertokens-auth-react/examples/with-passwordless:/usr/src/app \
```
Comment (rishabhpoddar) [inline examples/with-thirdpartypasswordless/scripts/start_container.sh:7]: Nood to change path to with-thirdpartypasswordless

File: examples/with-thirdpartypasswordless/scripts/start_container.sh [hunk 30]
```diff
@@ -0,0 +1,10 @@
+docker run -d \
+    --restart=always \
+    --name demo-app-passwordless \
```
Comment (rishabhpoddar) [inline examples/with-thirdpartypasswordless/scripts/start_container.sh:3]: Nood to change name to demo-app-thirdpartypasswordless

File: examples/with-thirdpartypasswordless/src/Home/SuccessView.js [hunk 31]
```diff
@@ -0,0 +1,52 @@
...
+            <div style={{ height: "10px" }} />
+            <div style={{ height: "10px" }} />
+            <div style={{ height: "10px" }} />
+            ------------------------------------
+            <div style={{ height: "10px" }} />
+            <div style={{ height: "10px" }} />
+            <div style={{ height: "10px" }} />
+            <div style={{ height: "10px" }} />
+            <a
+                href="https://github.com/supertokens/supertokens-auth-react/tree/master/examples/with-passwordless"
```
Comment (rishabhpoddar) [inline examples/with-thirdpartypasswordless/src/Home/SuccessView.js:45]: Change link to with-thirdpartypasswordless

File: lib/build/recipe/thirdpartypasswordless/types.d.ts (hunk 3 above)
Comment (rishabhpoddar) [reply lib/build/recipe/thirdpartypasswordless/types.d.ts:275]: Let's discuss this on a call.

File: lib/ts/version.ts [hunk 32]
```diff
@@ -12,5 +12,5 @@
  * License for the specific language governing permissions and limitations
  * under the License.
  */
-export const package_version = "0.19.0";
+export const package_version = "0.20.0";
 export const supported_fdi = ["1.8", "1.9", "1.10", "1.11", "1.12"];
```
Comment (rishabhpoddar) [inline lib/ts/version.ts:16]: Will need to add 1.13 here since that will have thirdpartypasswordless APIs

File: test/end-to-end/passwordless.test.js [hunk 33]
```diff
@@ -1388,17 +1401,17 @@ describe("SuperTokens Passwordless", function () {
                 assert.deepStrictEqual(consoleLogs, [
                     "ST_LOGS SESSION OVERRIDE ADD_FETCH_INTERCEPTORS_AND_RETURN_MODIFIED_FETCH",
                     "ST_LOGS SESSION OVERRIDE ADD_AXIOS_INTERCEPTORS",
-                    "ST_LOGS PASSWORDLESS OVERRIDE GET_LOGIN_ATTEMPT_INFO",
-                    "ST_LOGS PASSWORDLESS OVERRIDE CREATE_CODE",
-                    "ST_LOGS PASSWORDLESS PRE_API_HOOKS PASSWORDLESS_CREATE_CODE",
-                    "ST_LOGS PASSWORDLESS ON_HANDLE_EVENT PASSWORDLESS_CODE_SENT",
-                    "ST_LOGS PASSWORDLESS OVERRIDE SET_LOGIN_ATTEMPT_INFO",
+                    `ST_LOGS ${logId} OVERRIDE GET_LOGIN_ATTEMPT_INFO`,
```
Comment (rishabhpoddar) [inline test/end-to-end/passwordless.test.js:1404]: These changes are done cause all of these tests run for thirdpartypasswordless recipe as well?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: test/server/index.js [hunk 34]
```diff
@@ -60,6 +72,7 @@ let latestURLWithToken = "";
 
 let deviceStore = new Map();
 function saveCode({ email, phoneNumber, preAuthSessionId, urlWithLinkCode, userInputCode }) {
+    console.log(arguments[0]);
```
Comment (rishabhpoddar) [inline test/server/index.js:75]: Remove console log

File: test/server/index.js [hunk 35]
```diff
@@ -46,6 +46,18 @@ try {
...
+let thirdPartyPasswordlessSupported;
+let ThirdPartyPasswordlessRaw;
+let ThirdPartyPasswordless;
+
+try {
+    ThirdPartyPasswordlessRaw = require("supertokens-node/lib/build/recipe/thirdpartypasswordless/recipe").default;
+    ThirdPartyPasswordless = require("supertokens-node/recipe/thirdpartypasswordless");
+    thirdPartyPasswordlessSupported = true;
+} catch (ex) {
+    thirdPartyPasswordlessSupported = false;
```
Comment (rishabhpoddar) [inline test/server/index.js:58]: You should also change the `/test/featureFlags` API in this file to use this boolean?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: test/end-to-end/passwordless.test.js (hunk 33 above)
Comment (porcellus) [reply test/end-to-end/passwordless.test.js:1404]: yep.

File: test/server/index.js (hunk 34 above)
Comment (rishabhpoddar) [reply test/server/index.js:75]: We are keeping this - as discussed.

**Comment by rishabhpoddar** [comment]
@porcellus some of my previous comments are still pending. So moving this back to in progress.

**Review by rishabhpoddar — APPROVED** [review]
